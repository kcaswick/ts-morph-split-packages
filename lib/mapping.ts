import * as fs from "fs-extra";
import { ICruiseResult } from "dependency-cruiser";

export type IConfigJson = {
  OldPatterns: Record<string, Location>;
};

type IMadgeDependencies = Record<string, string[]>;

function isICruiseResult(o: unknown): o is ICruiseResult {
  return typeof o === "object" && o !== null && "modules" in o && "summary" in o;
}

export type ILocation = {
  Repo: string;
  Package: string;
  Path: string;
  Order?: number;
};

class Location implements ILocation {
  Repo: string;
  Package: string;
  Path: string;
  Order?: number | undefined;
  /**
   *
   */
  constructor(source: ILocation) {
    this.Repo = source.Repo ?? "";
    this.Package = source.Package ?? "";
    this.Path = source.Path ?? "";
    this.Order = source.Order ?? 0;
  }

  public toString = () =>
    `${this.Package === this.Repo ? "" : `"${this.Package}"`}${this.Repo}:${this.Path}`;
}

export type IMapResult = {
  OldName: string;
  New?: Location;
};

export class MapResult implements IMapResult {
  OldName: string;
  New?: Location;

  /**
   *
   */
  constructor(source: IMapResult) {
    this.OldName = source.OldName;
    this.New = source.New;
  }

  public isUnmapped = () => this.New === undefined;

  public toString = () => `${this.OldName}${this.isUnmapped() ? " (unmapped)" : ` => ${this.New}`}`;
}

export class Config {
  OldPatterns = new Map<string, Location>();
  BranchPrefix = "split/";
}

// TODO: Switch to using this class for mappings
export class Dependency {
  Name: MapResult;
  dependencyMap: MapResult[] = [];
  /**
   *
   */
  constructor({
    Name,
    dependencyMap: dependencies,
  }: {
    Name: MapResult;
    dependencyMap: MapResult[];
  }) {
    this.Name = Name;
    this.dependencyMap = dependencies;
  }

  public toString = (): string =>
    `${this.Name} has ${
      this.dependencyMap.length > 0 ? this.dependencyMap.length : "no "
    }dependencies [${this.dependencyMap.map((d) => d.toString()).join(", ")}]`;
}

export class PackageMapping {
  config = new Config();
  depMap: {
    Name: MapResult;
    dependencyMap: MapResult[];
  }[] = [];

  invdepMap: {
    Name: MapResult;
    dependentMap: MapResult[];
  }[] = [];

  public export() {
    let output = this.depMap.map(($_) => ({
      OldName: $_.Name.OldName,
      NewRepo: $_.Name.New?.Repo,
      NewPackage: $_.Name.New?.Package,
      NewName: $_.Name.New?.Path,
      "Dependency Count": $_.dependencyMap.length,
      "Package Dependencies": $_.dependencyMap
        .map(($_) => $_.New?.Package ?? "N/A")
        .sort()
        .filter((value, index, self) => self.indexOf(value) === index)
        .filter((value) => value !== $_.Name.New?.Package),
      dependencies: $_.dependencyMap.map(($_) =>
        undefined === $_.New ? $_.OldName : $_.New?.Repo + ":" + $_.New?.Path
      ),
    }));
    output = output.sort((a, b) =>
      (`${a.NewRepo}${a.NewName}` ?? a.OldName ?? "").localeCompare(
        `${b.NewRepo}${b.NewName}` ?? b.OldName
      )
    );

    return JSON.stringify(output, undefined, 2);
  }

  public exportPackageDependenciesChart(): IMadgeDependencies {
    const output: IMadgeDependencies = {};
    this.depMap.forEach((x) => {
      const packageName =
        (x.Name.New?.Package === "N/A" ? x.Name.New?.Repo : x.Name.New?.Package) ?? "N/A";
      output[packageName] ??= [];
      x.dependencyMap
        .map((d) => (d.New?.Package === "N/A" ? d.New?.Repo : d.New?.Package) ?? "(unmapped)")
        .filter((d) => d !== packageName)
        .filter(
          (value, index, self) =>
            self.indexOf(value) === index &&
            !output[packageName].find((already) => value === already)
        )
        .forEach((d) => output[packageName].push(d));
    });

    return output;
  }

  public inverseDependencies(dependencyList: Record<string, string[]>) {
    const dependents: Record<string, string[]> = {};
    Object.entries(dependencyList).forEach((p) => {
      p[1].forEach((_: string | number) => {
        dependents[_] ??= [];
        dependents[_].push(p[0]);
      });
    });
    return dependents;
  }

  public mapPackage(oldPath: string): Location | undefined {
    for (const oldPattern of this.config.OldPatterns.keys()) {
      let target: ILocation | undefined;
      if (
        new RegExp(oldPattern).test(oldPath) &&
        (target = this.config.OldPatterns.get(oldPattern))
      ) {
        const nw: Location = new Location(target);
        nw.Path = oldPath.replace(oldPattern, nw.Path ?? "");
        return nw;
      }
    }

    return undefined;
  }

  public getPackageMap(
    dependencyJsonPath = "doc/dependency.json",
    configPath = "./PackageMap.json"
  ) {
    const packageMap: IConfigJson = fs.readJsonSync(configPath);

    const dependenciesJson: IMadgeDependencies | ICruiseResult =
      fs.readJsonSync(dependencyJsonPath);

    const dep2: Record<string, string[]> = isICruiseResult(dependenciesJson)
      ? Object.fromEntries(
          dependenciesJson.modules
            .filter(
              (s) =>
                s &&
                !(
                  s.source.startsWith("node_modules") ||
                  s.source.startsWith("@") ||
                  s.source.endsWith(".css")
                )
            )
            .map((m) => [
              m.source,
              m.dependencies
                .map((d) => d.resolved)
                .filter(
                  (s) =>
                    s && !(s.startsWith("node_modules") || s.startsWith("@") || s.endsWith(".css"))
                ) ?? [],
            ])
            .values()
        )
      : dependenciesJson;

    const dependents = this.inverseDependencies(dep2);

    Object.entries(packageMap.OldPatterns)
      .sort((a, b) => (a[1].Order ?? 0) - (b[1].Order ?? 0))
      .forEach((x) => this.config.OldPatterns.set(...x));

    this.depMap = Object.entries(dep2).map((x) => ({
      Name: new MapResult({ OldName: x[0], New: this.mapPackage(x[0]) }),
      dependencyMap: x[1].map((y) => new MapResult({ OldName: y, New: this.mapPackage(y) })),
    }));

    this.invdepMap = Object.entries(dependents).map((x) => ({
      Name: new MapResult({ OldName: x[0], New: this.mapPackage(x[0]) }),
      dependentMap: x[1].map((y) => new MapResult({ OldName: y, New: this.mapPackage(y) })),
    }));

    return this;
  }
}
