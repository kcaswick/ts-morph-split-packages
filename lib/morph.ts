import { createProject, ts } from "@ts-morph/bootstrap";
import { basename } from "path";
import { ImportDeclaration } from "ts-morph";
import { getImportDeclarationsForSymbols } from "ts-morph-helpers";

import { PackageMapping } from "./mapping";

export async function prepareTsMorph(
  mapping: PackageMapping,
  currentRepo: string = basename(process.cwd())
) {
  // Read the existing project
  const project = await createProject({
    tsConfigFilePath: "tsconfig.json",
  });

  const languageService = project.getLanguageService();

  const sourceFiles = project.getSourceFiles();
  // Update imports in all programs based on the mapping
  sourceFiles
    .filter((sourceFile) => !sourceFile.fileName?.includes("node_modules"))
    .forEach((sourceFile) => {
      //   try {
      //     getImportDeclarationsForSymbols(
      //     languageService.getProgram().getTypeChecker().,
      //     sourceFile
      //   ).forEach((importDeclaration) => {
      //     const importPath = importDeclaration.getModuleSpecifierValue();
      //     const mappedPath = mapping.mapPackage(importPath);

      //     if (mappedPath) {
      //       if (mappedPath.Repo === currentRepo) {
      //         importDeclaration.setModuleSpecifier(mappedPath.Path);
      //       } else {
      //         importDeclaration.setModuleSpecifier(mappedPath.Package);
      //       }
      //     }
      //   });
      // } catch (e) {
      //   console.error(e);
      // }

      // try {
      sourceFile.forEachChild((node) => {
        if (ts.isImportDeclaration(node)) {
          const importPath = node.moduleSpecifier.getText();
          const mappedPath = mapping.mapPackage(importPath);

          console.debug(
            `${sourceFile.fileName}:${(node as any)?.getStartLineNumber?.()}:${node.getStart()}: ${node.getText()} => ${
              mappedPath ? node.getText().replace(importPath, mappedPath.Path) : ""
            } (${importPath})`
          );
          if (mappedPath) {
            if (mappedPath.Repo === currentRepo) {
              project.updateSourceFile(
                sourceFile.fileName,
                sourceFile.text.replace(importPath, mappedPath.Path)
              );
            } else {
              (node as any).setModuleSpecifier(mappedPath.Package);
            }
          }
        }
      });
      // } catch (e) {
      //   console.error(e);
      // }

      // try {
      //   (sourceFile as any)
      //     .getImportDeclarations()
      //     .forEach((importDeclaration: ImportDeclaration) => {
      //       const importPath = importDeclaration.getModuleSpecifierValue();
      //       const mappedPath = mapping.mapPackage(importPath);
      //       if (mappedPath) {
      //         if (mappedPath.Repo === currentRepo) {
      //           importDeclaration.setModuleSpecifier(mappedPath.Path);
      //         } else {
      //           importDeclaration.setModuleSpecifier(mappedPath.Package);
      //         }
      //       }
      //     });
      // } catch (e) {
      //   console.error(e);
      // }

      // try {
      //   sourceFile.statements.forEach((statement) => {
      //     if (ts.isImportDeclaration(statement)) {
      //       const importPath = (statement as any).getModuleSpecifierValue();
      //       const mappedPath = mapping.mapPackage(importPath);
      //       if (mappedPath) {
      //         if (mappedPath.Repo === currentRepo) {
      //           (statement as any).setModuleSpecifier(mappedPath.Path);
      //         } else {
      //           (statement as any).setModuleSpecifier(mappedPath.Package);
      //         }
      //       }
      //     }
      //   });
      // } catch (e) {
      //   console.error(e);
      // }

      // try {
      //   (sourceFile as any).imports.forEach((importDeclaration: ImportDeclaration) => {
      //     const importPath = importDeclaration.getModuleSpecifierValue();
      //     const mappedPath = mapping.mapPackage(importPath);
      //     if (mappedPath) {
      //       if (mappedPath.Repo === currentRepo) {
      //         importDeclaration.setModuleSpecifier(mappedPath.Path);
      //       } else {
      //         importDeclaration.setModuleSpecifier(mappedPath.Package);
      //       }
      //     }
      //   });
      // } catch (e) {
      //   console.error(e);
      // }
    });
}
