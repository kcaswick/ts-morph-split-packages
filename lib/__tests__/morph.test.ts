import path from "path";
import { ls, popd, pushd } from "shelljs";
import simpleGit, { CheckRepoActions, SimpleGit } from "simple-git";
import { Project } from "ts-morph";

import { MapResult, PackageMapping } from "../mapping";
import * as morph from "../morph";
import {
  advanceToPhase,
  checkoutTempSimpleRepo,
  expectImportChanged,
  getInternalImportsFlat,
  IMapPhaseState,
  importNodeToText,
  loadSimpleMadge,
  ProcessPhase,
} from "./test_fixtures";

async function arrangeRepoAndSnapshot(
  branch?: string
): Promise<[string, SimpleGit, IMapPhaseState]> {
  const [tempRepoPath, tempRepo] = await checkoutTempSimpleRepo();
  const testState = await advanceToPhase(
    ProcessPhase.Map,
    { currentPhase: ProcessPhase.Initial, tempRepoPath, tempRepo },
    loadSimpleMadge
  );
  if (branch) {
    await tempRepo.checkout(branch);
  } else {
    await tempRepo.checkoutLocalBranch("test_branch");
  }

  const status = await testState.tempRepo.status();
  expect(status).toMatchSnapshot("tempRepo.status");
  return [tempRepoPath, tempRepo, testState];
}

describe("test morph", function () {
  it("test morph.prepareTsMorph", async function () {
    const mapping = loadSimpleMadge();
    const modifiedProject = await morph.prepareTsMorph(mapping);
    expect(modifiedProject).toBeDefined();
    console.debug("Start snapshot");
  });
  test("prepareTsMorph simple repo", async () => {
    // Arrange
    const [tempRepoPath, tempRepo, testState] = await arrangeRepoAndSnapshot();
    // TODO: Figure out if the right branch is checked out, or if we need to switch

    // Act
    pushd(tempRepoPath);
    const modifiedProject = await morph.prepareTsMorph(testState.packageMapping);
    popd();
    expect(modifiedProject).toBeDefined();

    // Assert

    // Make sure it isn't doing:
    //    C:/Users/kcaswick/source/repos/External/ts-morph-split-packages/lib/__tests__/morph.test.ts:…:
    //    import {…} from "./test_fixtures"; =>  ("./test_fixtures")
    // instead:
    //    import {…} from "./test_fixtures"; => ("../../test_fixtures")
    const importDeclarationsFlat = getInternalImportsFlat(modifiedProject.project);
    const importDeclarationsFlatText = importDeclarationsFlat
      .map(importNodeToText(tempRepoPath))
      .sort();
    expect(importDeclarationsFlatText).toMatchSnapshot("importDeclarationsFlatText");
    expectImportChanged(importDeclarationsFlatText, "test_fixtures", '"./test_fixtures');
  }, 15000);
  test("prepareTsMorph simple repo test_fixtures pkg", async () => {
    // Arrange
    const [tempRepoPath, _tempRepo, mapState] = await arrangeRepoAndSnapshot();

    // Act
    pushd(tempRepoPath);
    const rewriteState = await advanceToPhase(ProcessPhase.Rewrite, mapState, loadSimpleMadge);

    // Filter to just the files in the test_fixtures package
    const moveState = await advanceToPhase(ProcessPhase.Move, rewriteState);
    await mapState.tempRepo.checkout("split/test_fixtures");
    const finalProject = new Project({
      tsConfigFilePath: path.join(tempRepoPath, "tsconfig.json"),
    });
    popd();

    // Assert
    const importDeclarationsFlat = getInternalImportsFlat(finalProject);
    const importDeclarationsFlatText = importDeclarationsFlat
      .map(importNodeToText(tempRepoPath))
      .sort();
    expect(importDeclarationsFlatText).toMatchSnapshot("importDeclarationsFlatText");
    expectImportChanged(importDeclarationsFlatText, "PackageMapping", "..");
  }, 15000);
  test("save modified simple repo", async () => {
    // Arrange
    const [tempRepoPath, tempRepo] = await checkoutTempSimpleRepo();
    const testState = await advanceToPhase(
      ProcessPhase.Map,
      { currentPhase: ProcessPhase.Initial, tempRepoPath, tempRepo },
      loadSimpleMadge
    );
    expect(await tempRepo.status()).toMatchSnapshot(`tempRepo.status before`);
    pushd(tempRepoPath);
    try {
      const targetRepos = testState.packageMapping.depMap
        .flatMap((x) => (x.Name.isMapped() ? x.Name.New.Repo : []))
        .filter((value, index, self) => self.indexOf(value) === index)
        .sort();
      expect(targetRepos).toMatchInlineSnapshot(`
        Array [
          "new",
          "test_fixtures",
          "ts-morph-split-packages",
        ]
      `);
      // testState.packageMapping.depMap
      // .filter(
      //   (x) =>
      //     x.Name.isMapped() &&
      //     (x.Name.New.Repo === "N/A" || x.Name.New.Repo === baseRepoName) &&
      //     x.Name.OldName !== x.Name.New.Path
      // )
      // .map((x) => [x.Name.OldName, x.Name.New?.Path ?? ""]);
      await Promise.allSettled(
        targetRepos.map(async (targetRepo) => {
          console.debug(
            await testState.tempRepo.checkout(
              testState.packageMapping.config.BranchPrefix + targetRepo
          ));

          const { project: modifiedProject, modifiedFiles: modifiedFilesSet } =
            await morph.prepareTsMorph(testState.packageMapping, targetRepo);
          console.debug(
            `modifiedFilesSet.size for '${targetRepo}': ${modifiedFilesSet.size}`,
            modifiedFilesSet
          );
          if (modifiedFilesSet.size === 0) {
            return;
          }

          // Act
          await modifiedProject.save();

          // Assert
          const status = await testState.tempRepo.status();
          expect(status).toMatchSnapshot(`tempRepo.status '${targetRepo}' after`);

          await testState.tempRepo.commit("Rewrite imports", ["--all"]);

          // Get the list of file names that were modified
          const modifiedFileNames = Array.from(modifiedFilesSet.values()).map((x) => x.getFilePath());

          // If and only if a file is in the list, it should be modified
          expect(status.modified.sort()).toContainEqual(modifiedFileNames.sort());
        })
      );
    } finally {
      popd();
    }
  }, 15000);
});
