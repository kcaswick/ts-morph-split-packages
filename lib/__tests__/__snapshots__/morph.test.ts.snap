// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test morph prepareTsMorph simple repo main pkg: importDeclarationsFlatText 1`] = `
Array [
  "lib\\\\__tests__\\\\complexRepo.test.ts: import { createWriteStream /* , mkdirSync */ } from \\"fs\\";",
  "lib\\\\__tests__\\\\git.test.ts: import { loadSimpleMadge } from \\"test_fixtures\\";",
  "lib\\\\__tests__\\\\tsMorphSplitPackages.test.ts: import { createWriteStream } from \\"fs\\";",
  "lib\\\\git.ts: import { basename } from \\"path\\";",
]
`;

exports[`test morph prepareTsMorph simple repo main pkg: tempRepo.status 1`] = `
StatusSummary {
  "ahead": 0,
  "behind": 0,
  "conflicted": Array [],
  "created": Array [],
  "current": "split/ts-morph-split-packages",
  "deleted": Array [],
  "detached": false,
  "files": Array [],
  "ignored": undefined,
  "isClean": [Function],
  "modified": Array [],
  "not_added": Array [],
  "renamed": Array [],
  "staged": Array [],
  "tracking": null,
}
`;

exports[`test morph prepareTsMorph simple repo main pkg: test_fixtures 1`] = `
Array [
  "lib\\\\__tests__\\\\git.test.ts: import { loadSimpleMadge } from \\"test_fixtures\\";",
]
`;

exports[`test morph prepareTsMorph simple repo test_fixtures pkg: PackageMapping 1`] = `
Array [
  "lib\\\\package\\\\test_fixtures\\\\__tests__\\\\test_fixtures.ts: import { PackageMapping } from \\"new\\";",
]
`;

exports[`test morph prepareTsMorph simple repo test_fixtures pkg: importDeclarationsFlatText 1`] = `
Array [
  "lib\\\\package\\\\test_fixtures\\\\__tests__\\\\test_fixtures.ts: import { PackageMapping } from \\"new\\";",
]
`;

exports[`test morph prepareTsMorph simple repo test_fixtures pkg: tempRepo.status 1`] = `
StatusSummary {
  "ahead": 0,
  "behind": 0,
  "conflicted": Array [],
  "created": Array [],
  "current": "split/test_fixtures",
  "deleted": Array [],
  "detached": false,
  "files": Array [],
  "ignored": undefined,
  "isClean": [Function],
  "modified": Array [],
  "not_added": Array [],
  "renamed": Array [],
  "staged": Array [],
  "tracking": null,
}
`;

exports[`test morph save modified simple repo: tempRepo.status before 1`] = `
StatusSummary {
  "ahead": 0,
  "behind": 0,
  "conflicted": Array [],
  "created": Array [],
  "current": "split/ts-morph-split-packages",
  "deleted": Array [],
  "detached": false,
  "files": Array [],
  "ignored": undefined,
  "isClean": [Function],
  "modified": Array [],
  "not_added": Array [],
  "renamed": Array [],
  "staged": Array [],
  "tracking": null,
}
`;
